design/open/kv-query-tool.md

```diff
@@ -14,32 +16,69 @@ We will elevate this logic and into a new command that will provide the ability
 lakefs kv <cmd> <args>
 ```
 
-### Get/Scan/Set
+## Implementation Steps
+
+These steps describe the order and importance of the functionality we would like to get from `lakefs kv` and does not refer to specific commands format. The commands specified are for illustration purposes. The specifics of the commands format and their arguments are discussed in later section
+
+### Step 1 - Get/Scan
+
+* `lakefs kv get <args>`
+* `lakefs kv scan <args>`
+
+These 2 are `lakefs kv`'s MVP. They put the **query** in "**query**ing tool". Given some kind of identifier, these commands returns an object, or scans through a range of objects, outputting their details. The identifier can be a (`partition`,`key`) pair, or an object-type and its identifier (as used by lakefs, e.g. RepositoryID, BranchID etc.), depends on the approach we select - to be discussed later
+
+### Step 2 - REPL
+
+An interactive CLI prompt
+```
+$ lakefs kv repl
+
+kv>
+```
+The possible commands are no other than the ones supported by `lakefs kv`, i.e. Get & Scan, and if step 3 is to be implemented, Set and Delete
+
+The benefits of this approach is that unlike the `lakefs kv XXX` (where XXX is `get` or `set`), which runs as a dedicated process and terminates once input is generated, the `repl` command can execute multiple commands in the same process, saving the need to recreate data structures and connect to KV, as well as reducing the chance of failures due to errors in the above
+
+### Step 3 - Set/Delete - If the need arises
+
+* `lakefs kv set <args>`
+* `lakefs kv delete <args>`
+
+These commands can be added in case the need arises. The first is to alter/add a KV store entry, and the other is for removing an entry. These are very error prone operations that can easily result in data corruption, and so we will (a) not implement these right away, but only if they are really needed and (b) protect these with, at least, a requirement for command confirmation: `"Are you sure you want to delete entry? [yes/N]" etc.
+
+## Different Approaches, Different Interfaces, Different Problems
+
+### Full Knowledge
+
+With this approach, `lakefs kv` is implemented with all the knowledge available. Knowledge like: how to generate a repository partition from the repository ID and object, how to generate a branch key from the repository and branch IDs, the fact that staged objects' keys cannot be provided as strings etc.
+The commands in that case will be very specific, having the user specifying which entity he is working with. Essentially, this will be a CLI interface to various services, using KV (graveler, auth, etc.)
+* `lakefs kv get-repository <REPO_ID>`
```

Comment:
How is it different than `lakectl repo get <REPO_ID>`?

Reply:
First, `lakectl repo get` involves an API call, but I don't think that what you meant.
`lakectl repo get` will get the repository from graveler's ref manager (via catalog, via controller, via API ofcourse), and will get it processed and chewed. For instance, if the repository state is `RepositoryState_IN_DELETION`, `lakectl repo get` will get an error rather than a repositry. This is not my intention for `lakefs kv` - it will return the repository as it extracted from the KV, with the minimal processing of unmarshalling. That way, `lkaefs kv` will be useful in debugging KV issues, as it will not be amsked by other logic. Does it make sense?